### ISSUE TITLE
Jest --coverage : help

### ISSUE THREAD

**Issue by seeliang** [nickmerwin/node-coveralls#126]
Hi everyone,

is anyway i can setup this with Jest?

many thanks

**Comment by jseminck** [comment]
Just add: `&& cat ./coverage/lcov.info | ./node_modules/coveralls/bin/coveralls.js`

```js
    "test": "node --harmony_proxies node_modules/.bin/jest --coverage && cat ./coverage/lcov.info | ./node_modules/coveralls/bin/coveralls.js",
```

**Comment by tizmagik** [comment]
Thanks @jseminck would be good to add this to the README, I think?

**Comment by ndelvalle** [comment]
I follow @jseminck approach, but now Im getting the folling error:
```
[error] "2016-11-04T14:51:39.503Z"  'error from lcovParse: ' 'Failed to parse string'
[error] "2016-11-04T14:51:39.505Z"  'input: ' ''
[error] "2016-11-04T14:51:39.506Z"  'error from convertLcovToCoveralls'
```
Any ideas? Thanks in advanced.

**Comment by cdaringe** [comment]
@ndelvalle, most likely, your `coverage/lcov.info` is empty

**Comment by ndelvalle** [comment]
yep @cdaringe that was the problem 👍

**Comment by joseluisq** [comment]
:+1:

**Comment by bcoe** [comment]
@jseminck @seeliang worth noting that Jest's coverage implementation outputs a .JSON report that is compatible with nyc and Istanbul in the `coverage/` folder. So this would totally work:

`nyc report --temp-directory=coverage --reporter=text-lcov | coveralls`

**Comment by seeliang** [comment]
Thanks, guys.
I will give a go later :)

**Comment by iddan** [comment]
Can this be added to the README?

**Comment by nickmccurdy** [comment]
👍 For adding to readme.

@bcoe That's cool, are you aware of any advantage to using nyc's JSON output over `coveralls < coverage/lcov.info`? As far as I'm aware I don't think there's a difference in the formats given to coveralls and it's convenient to not have to install nyc.

**Comment by arkgast** [comment]
Why not use a script like this?

    jest --coverage && cat ./coverage/lcov.info | coveralls

**Comment by nickmccurdy** [comment]
Personally I like to set `"jest": { "collectCoverage": true }` in `package.json` to enable coverage reporting everywhere, add coveralls as a `devDependency`, and have CI run `coveralls < coverage/lcov.info`.

**Comment by arkgast** [comment]
😮 now I got it, so most people don't have `coveralls` as a `devDependency` that's way `./node_modules/coveralls/bin/coveralls.js` appears in their configurations.

I was asking about that part specially.

By the way I think this is the best approach to maintain configurations clearer. As a side effect you had clarify my doubt 😆, thanks man

**Comment by nickmccurdy** [comment]
I don't think that's the case. Assuming that the other technique is to `npm install coveralls` in CI after a build, it would still install the package to `node_modules` with a binary in that location. Whether or not you install as a saved `devDependency`, it will only be executable directly via `coveralls` if you have it in your `$PATH` (some CI services like Travis do this automatically) or if you're calling it in an npm script (which does something similar).

**Comment by dennisja** [comment]
@nickmccurdy Your solution isnot supported by create-react-app
It gives the warning below:
Out of the box, Create React App only supports overriding these Jest options:
  • collectCoverageFrom
  • coverageReporters
  • coverageThreshold
  • snapshotSerializers.
These options in your package.json Jest configuration are not currently supported by Create React App:
  • collectCoverage

**Comment by nickmccurdy** [comment]
Yes, but you can still use `jest --coverage` in your build script even though the config option isn't supported.

**Comment by dennisja** [comment]
Yeah,
I solved it by run `yarn test --coverage` which later runs `jest --coverage`

**Comment by ummahusla** [comment]
This is how I got it to work

```
node_modules/.bin/jest --coverage && cat ./coverage/lcov.info | ./node_modules/coveralls/bin/coveralls.js
```

**Comment by jeff1evesque** [comment]
@jseminck, @tizmagik: sorry to bother in this thread. I'm using jest + enzyme, and hoping to pipe the coverage results to coveralls.  So, I implemented `test:coveralls`, a `script` from my [`package.json`](https://github.com/jeff1evesque/machine-learning/blob/6f1be85c28ae688825a13e5e512ecd42177c6ca7/src/jsx/package.json#L10-L11):

```json
  "scripts": {
    "build:browserify": "browserify -t [ babelify --presets env,stage-2,react ] content.jsx > ../../interface/static/js/content.js",
    "prebuild:dos2unix": "find . -type f -print0 | xargs -0 dos2unix",
    "watch:jsx": "onchange 'src/jsx/**/*.jsx' 'src/jsx/**/*.js' -- npm run build:browserify",
    "test": "NODE_ENV=test jest --config jest.config.js",
    "test:coveralls": "npm run test --coverage --coverageReporters=lcov.info | ./node_modules/coveralls/bin/coveralls.js"
  },
```

I've [verified](https://github.com/jeff1evesque/machine-learning/pull/3235#issuecomment-393588327) that `lcov.info`, should be nonempty. However, my travis [build](https://travis-ci.org/jeff1evesque/machine-learning/jobs/386263115#L499-L517), seems to error:

```bash
[error] "2018-05-31T16:37:12.946Z"  'error from convertLcovToCoveralls'
/home/travis/build/jeff1evesque/machine-learning/src/jsx/node_modules/coveralls/bin/coveralls.js:18
        throw err;
        ^
Failed to parse string
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! reactjs@1.0.0 test:coveralls: `npm run test --coverage --coverageReporters=lcov.info | ./node_modules/coveralls/bin/coveralls.js`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the reactjs@1.0.0 test:coveralls script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
npm ERR! A complete log of this run can be found in:
npm ERR!     /home/travis/.npm/_logs/2018-05-31T16_37_12_965Z-debug.log
The command "npm run test:coveralls" exited with 1.
```

**Comment by jeff1evesque** [comment]
@jseminck, @tizmagik: nevermind, I think I've fixed it. The corresponding travis [build](https://travis-ci.org/jeff1evesque/machine-learning/jobs/386322980#L593-L594), seems to have run the following without erroring:

```bash
cat lcov.info | ./node_modules/coveralls/bin/coveralls.js
```

I guess I'll merge my corresponding [pull request](https://github.com/jeff1evesque/machine-learning/pull/3235), and see if my [`webhook`](https://github.com/jeff1evesque/machine-learning/blob/8703804945703f301d50297df2abaac661bc083e/.travis.yml#L19-L20) updates coveralls.

**Comment by jeff1evesque** [comment]
Ah, didn't have to wait for PR to be merged. Seems like the configurations [worked](https://coveralls.io/builds/17257835). Thank you everyone, pretty much recycled different aspects from various comments in this issue.

**Comment by petermikitsh** [comment]
@nickmccurdy's solution worked for me. Thanks Nick!

**Comment by ambroiseRabier** [comment]
Based on @jeff1evesque and @nickmccurdy answer:

*package.json*
```json
"scripts": {
  "ci-process-cover": "cat coverage/lcov.info | coveralls"
}
```

*jest.config.js*
```js
module.exports = {
  "collectCoverage": true
  // ...
}
```

*.travis.yml*
```yml
script:
  - npm run test
after_script:
  - npm run ci-process-cover
```

Or have tests run twice:

*package.json*
```
`cover`: `jest --coverage --coverageReporters=text-lcov | coveralls`
```

*.travis.yml*
```yml
script:
  - npm run test
after_script:
  -  npm run cover
```

If you replace `npm run test` by `npm run cover`, coverage will work, but your CI will not fail if one test fail.

**Comment by zdrawku** [comment]
I've faced a similar problem, hope [this post](https://stackoverflow.com/questions/60362121/how-to-configure-coveralls-with-github-action/65615948#65615948) might help you a little bit.

**Comment by adefemi171** [comment]
Hi all am facing 
```
[error] "2021-01-11T05:26:15.227Z"  'error from lcovParse: ' 'Failed to parse string'
[error] "2021-01-11T05:26:15.230Z"  'input: ' ''
[error] "2021-01-11T05:26:15.230Z"  'error from convertLcovToCoveralls'
```
 when running it locally and in my Github action. How do I generate .lcov so the file won't be empty? @nickmccurdy @jseminck @tizmagik

**Comment by zdrawku** [comment]
@adefemi171 if you are using Angular [`--code-coverage` ](https://angular.io/cli/test)will generate a coverage report for you.

**Comment by adefemi171** [comment]
Am using Node and Jest and I ran `jest --coverage` which created a coverage folder but with an empty `lcov.info` @zdrawku

**Comment by Yegorich555** [comment]
> Why not use a script like this?
> 
> ```
> jest --coverage && cat ./coverage/lcov.info | coveralls
> ```

Because it doesn't work on Windows. 

Do we have any cross-platform solution ?

I figured only following:
```js
jest --coverage && coveralls < coverage\\lcov.info
```
Ordinary pipe `jest --coverage --coverageReporters=text-lcov | coveralls` somehow doesn't work with NodeJS 14.15.*
